libgnucash/engine/ScrubBudget.c

```diff
@@ -0,0 +1,95 @@
+/********************************************************************\
+ * Scrub.c -- convert single-entry accounts into clean double-entry *
+ *                                                                  *
+ * This program is free software; you can redistribute it and/or    *
+ * modify it under the terms of the GNU General Public License as   *
+ * published by the Free Software Foundation; either version 2 of   *
+ * the License, or (at your option) any later version.              *
+ *                                                                  *
+ * This program is distributed in the hope that it will be useful,  *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of   *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
+ * GNU General Public License for more details.                     *
+ *                                                                  *
+ * You should have received a copy of the GNU General Public License*
+ * along with this program; if not, contact:                        *
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+ *                                                                  *
+\********************************************************************/
+
+#include <config.h>
+
+
+#include <glib.h>
+#include <glib/gi18n.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <libguile.h>
+
+#include "ScrubBudget.h"
+#include "gnc-budget.h"
+#include "gnc-features.h"
+#include "gnc-prefs.h"
+#include "guid.h"
+/* #include "Account.h" */
+/* #include "AccountP.h" */
+/* #include "Scrub.h" */
+/* #include "ScrubP.h" */
+/* #include "gnc-commodity.h" */
+/* #include "qofinstance-p.h" */
+
+/* ================================================================ */
+static char* reverse_budget_scm =
+    "(lambda (budget-guid) \
+    (let* ((book (gnc-get-current-book)) \
+           (budget (gnc-budget-lookup budget-guid book)) \
+           (accounts (gnc-account-get-descendants (gnc-book-get-root-account book))) \
+           (periods (iota (gnc-budget-get-num-periods budget)))) \
+      (for-each \
+       (lambda (account) \
+         (for-each \
+          (lambda (period) \
+            (let ((amt (gnc-budget-get-account-period-value budget account period))) \
+              (gnc-budget-set-account-period-value budget account period (- amt)))) \
+          periods)) \
+       (filter gnc-reverse-balance accounts))))";
```

Comment:
C-calling-scheme, please. Based on the names all of those functions are wrapped from C anyway.

Reply:
Baby steps! The previous commit would load normal scheme in gnome-utils.scm. We'll want to move scrub to libgnucash which incidentally mean there'll be no more dialog to offer conversion.

Reply:
Rewrite-to-C *then* move it to `engine`.